src/base-subscription.ts

```diff
@@ -1,252 +1,37 @@
-import { SubscriptionEvent } from './subscription';
+import { SubscriptionEvent, Subscription, SubscriptionTransport } from './subscription';
 
-export enum BaseSubscriptionState {
-    UNOPENED = 0, // haven't called xhr.send()
-    OPENING,      // called xhr.send(); not yet received response headers
-    OPEN,         // received response headers; called onOpen(); expecting message
-    ENDING,       // received EOS message; response not yet finished
-    ENDED         // called onEnd() or onError(err)
-}
-
-export type BaseSubscriptionConstruction = (headers: ElementsHeaders) => Promise<BaseSubscription>;
-        
-    export class BaseSubscription {
-        
-        private state: BaseSubscriptionState = BaseSubscriptionState.UNOPENED;
-        
-        constructor(
-            private xhr: XMLHttpRequest,
-            private logger: Logger,
-            private onOpen: (headers: ElementsHeaders) => void = headers => {},
-            private onError: (error: any) => void = error => {}, 
-            private onEvent: (event: SubscriptionEvent) => void = event => {},
-            private onEnd: (error?: any) => void = error => {}
-        ){            
-            xhr.onreadystatechange = () => {
-                switch(this.xhr.readyState) {
-                    case XhrReadyState.UNSENT:
-                    case XhrReadyState.OPENED:
-                    case XhrReadyState.HEADERS_RECEIVED:
-                    this.assertStateIsIn(BaseSubscriptionState.OPENING);
-                    break;
-                    
-                    case XhrReadyState.LOADING:
-                    this.onLoading();
-                    break;
-                    
-                    case XhrReadyState.DONE:
-                    this.onDone();
-                    break;
-                }
-            }
-            this.state = BaseSubscriptionState.OPENING;
-            this.xhr.send();
-        }  
-        
-        public unsubscribe(): void {
-            this.state = BaseSubscriptionState.ENDED;
-            this.xhr.abort();
-            this.onEnd();
-        }
+export class BaseSubscription implements Subscription {
+    private subID: number;
+    
+    constructor(
+        private path: string,
+        private transport: SubscriptionTransport,
+        private headers: ElementsHeaders,
+        private logger: Logger,
+        private onOpen: (headers: ElementsHeaders) => void = headers => {},
+        private onError: (error: any) => void = error => {}, 
+        private onEvent: (event: SubscriptionEvent) => void = event => {},
+        private onEnd: (error?: any) => void = error => {},
+        subID?: number
+    ) {
+        this.subID = this.transport.subscribe(
+            path,
+            {
+                onOpen,
+                onError,
+                onEvent,
+                onEnd
+            },
+            headers,
+            subID
+        );
+    }
```

Comment:
Why does it return an id instead of a subscription? Is it because of cancellation?

Reply:
It does not return an id. It's just hold for cancellation as you mentioned.

`
public unsubscribe () {
   this.transport.unsubscribe(this.subID);
 }
`

Reply:
I posted this comment in a wrong place. I was talking about the call above - `this.transport.subscribe` returning a subscription id. Sorry for confusion.